clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -471,10 +547,31 @@ private void handleDisconnections(List<ClientResponse> responses, long now) {
     /**
      * Record any newly completed connections
      */
-    private void handleConnections() {
+    private void handleConnections(long now) {
         for (String node : this.selector.connected()) {
             log.debug("Completed connection to node {}", node);
             this.connectionStates.connected(node);
```

Comment:
Apologies for the late review, but I feel like it would be much clearer if the state transition were:
`DISCONNECTED -> CONNECTING -> CHECKING_API_VERSIONS -> CONNECTED`

It strikes me that the main reason you have the READY state is to manage `canSendMore`, but I can't actually find where `canSendMore` is called in the api version request flow. NetworkClient `doSend` skips the `canSendMore` check, I believe, so do we really need that extra layer of complexity?

And so would it be possible to check for `this.requiredApiVersions == null` directly here and set the next connectionState to CONNECTED if null, or to CHECKING_API_VERSIONS if not? Then in the api version response handler the state could be updated to CONNECTED.